include/jni/arraylike.hpp

```diff
@@ -30,8 +30,11 @@ namespace jni
     template < class E, std::size_t n >    E const * ArraylikeData(const std::array<E,n>& a)              { return a.data(); }
     template < class E, class A >          E       * ArraylikeData(      std::vector<E,A>& a)             { return a.data(); }
     template < class E, class A >          E const * ArraylikeData(const std::vector<E,A>& a)             { return a.data(); }
-    template < class C, class T, class A > C       * ArraylikeData(      std::basic_string<C,T,A>& a)     { return &a[0]; }
-    template < class C, class T, class A > C const * ArraylikeData(const std::basic_string<C,T,A>& a)     { return &a[0]; }
+    template < class C, class T, class A > C       * ArraylikeData(      std::basic_string<C,T,A>& a)     { return a.data(); }
+    template < class C, class T, class A > C const * ArraylikeData(const std::basic_string<C,T,A>& a)     { return a.data(); }
+    // For supporting std::string <-> jbyte conversions
+    template < class T, class A >      jbyte       * ArraylikeData(      std::basic_string<char,T,A>& a)     { return (jbyte*)a.data(); }
```

Comment:
These overloads overlap with the two templated overloads `ArraylikeData` above. Since they're non-templated, they'll take precedence: if you pass `ArraylikeData` a `std::string`, the compiler will select the `jbyte*`-returning version, which is not always what you want. So instead of adding new `ArraylikeData` overloads, I think you'll need to adjust the definition of the added `MakeAnything` overloads to include the cast. Also, it should be a `reinterpret_cast` rather than a C-style cast. The latter are prohibited in jni.hpp. (You can run `make test` to make sure your additions compile.)

Reply:
Makes sense.